dimod/discrete/discrete_quadratic_model.py

```diff
@@ -217,6 +214,104 @@ def add_linear_equality_constraint(self, terms: LinearTriplets,
         self._cydqm.add_linear_equality_constraint(
             index_terms, lagrange_multiplier, constant)
 
+    def add_linear_inequality_constraint(self, terms: LinearTriplets,
+                                         lagrange_multiplier: float,
+                                         label: str,
+                                         constant: int = 0,
+                                         lb: int = - np.Infinity,
+                                         ub: int = 0,
+                                         slack_method: str = "log2",
+                                         cross_zero: bool = False):
+
+        """Add a linear inequality constraint as a quadratic objective.
+
+        Adds a linear inequality constraint of the form:
+
+        math:'lb <= \sum_{i,k} a_{i,k} x_{i,k} + constant <= ub'
+        to the discrete quadratic model as a quadratic objective. Coefficients should be integers.
+        For constraints with fractional coefficients, multiply both sides of the inequality by an
+        appropriate factor of ten to attain or approximate integer coefficients.
+
+        Args:
+            terms:
+                A list of tuples of the type (variable, case, bias).
+                Each tuple is evaluated to the term (bias * variable_case).
+                All terms in the list are summed.
+            lagrange_multiplier:
+                The coefficient or the penalty strength
+            label:
+                Prefix used to label the slack variables used to create the new objective.
+            constant:
+                The constant value of the constraint.
+            lb:
+                lower bound for the constraint
+            ub:
+            slack_method:
+                "The method for adding slack variables. Supported methods are:
+                - log2: Adds log2(slack_range) number of dqm variables each with two cases to the constraint.
+                - log10: Adds log10 dqm variables each with up to 10 cases.
+
+                upper bound for the constraint
+            cross_zero
+                If True, includes zero as the domain of constraint if not include already
+       """
+        assert (slack_method in ['log2', 'log10'])
+
+        terms_upper_bound = sum(v for _, _, v in terms if v > 0)
```

Comment:
Since we are passing through the terms more than once, we actually require terms to be a `Collection`. Right now I could pass in a generator and this would fail since the generator would be exhausted by this line.

Example:
```
>>> terms_list = [('a', 'b', 1), ('b', 'c', 2), ('c', 'd', -1)]
>>> print(sum(v for _, _, v in terms_list if v > 0))
3
>>> print(sum(v for _, _, v in terms if v < 0))
-1
```
works fine, but 
```
>>> terms_generator = (t for t in [('a', 'b', 1), ('b', 'c', 2), ('c', 'd', -1)])
>>> print(sum(v for _, _, v in terms_generator if v > 0))
3
>>> print(sum(v for _, _, v in terms_generator if v < 0))
0
```
both `terms_list` and `terms_generator` are `Iterable`.

An alternative would be something like

```
if isinstance(terms, Iterator):
    terms = list(terms)